packages/loader/container-loader/src/connectionStateHandler.ts

```diff
@@ -179,43 +212,62 @@ export class ConnectionStateHandler {
         details: IConnectionDetails,
     ) {
         const oldState = this._connectionState;
         this._connectionState = ConnectionState.CatchingUp;
+        const writeConnection = connectionMode === "write";
+
+        // Note that this may be undefined since the connection is established proactively on load
+        // and the quorum may still be under initialization.
+        const quorumClients: IQuorumClients | undefined = this.handler.quorumClients();
 
         // Stash the clientID to detect when transitioning from connecting (socket.io channel open) to connected
         // (have received the join message for the client ID)
         // This is especially important in the reconnect case. It's possible there could be outstanding
         // ops sent by this client, so we should keep the old client id until we see our own client's
-        // join message. after we see the join message for out new connection with our new client id,
+        // join message. after we see the join message for our new connection with our new client id,
         // we know there can no longer be outstanding ops that we sent with the previous client id.
         this._pendingClientId = details.clientId;
 
-        // Report telemetry after we set client id, but before transitioning to Connected state below!
+        // This pending clientId could be in the quorum already (i.e. join op already processed).
+        // We are fetching ops from storage in parallel to connecting to Relay Service,
+        // and given async processes, it's possible that we have already processed our own join message before
+        // connection was fully established.
+        // If quorumClients itself is undefined, we expect it will process the join op after it's initialized.
+        const waitingForJoinOp = quorumClients?.getMember(this._pendingClientId) === undefined;
+
+        // IMPORTANT: Report telemetry after we set _pendingClientId, but before transitioning to Connected state
         this.handler.logConnectionStateChangeTelemetry(ConnectionState.CatchingUp, oldState);
 
-        const quorumClients = this.handler.quorumClients();
-        // Check if we already processed our own join op through delta storage!
-        // we are fetching ops from storage in parallel to connecting to ordering service
-        // Given async processes, it's possible that we have already processed our own join message before
-        // connection was fully established.
-        // Note that we might be still initializing quorum - connection is established proactively on load!
-        if (quorumClients?.getMember(details.clientId) !== undefined
-            || connectionMode === "read"
-        ) {
-            assert(!this.prevClientLeftTimer.hasTimer, 0x2a6 /* "there should be no timer for 'read' connections" */);
-            this.setConnectionState(ConnectionState.Connected);
-        } else if (connectionMode === "write") {
+        if (writeConnection && waitingForJoinOp) {
```

Comment:
Correct me if I got it wrong, but all the logic stayed the same with exception when assert fires.
It used to be equivalent to this assert:
    assert(waitingForJoinOp && writeConnection || !this.waitingForLeaveOp);
and it became 
    assert(writeConnection || !this.waitingForLeaveOp);

Reply:
Can we move discussion to https://github.com/microsoft/FluidFramework/pull/10570 please?

Reply:
Not quite.  When I first fixed the assert, the test still failed.  If we are `waitingForLeaveOp` we must not transition to `Connected` here, but rather do nothing.